subsys/zigbee/osif/zb_nrf_serial.c

```diff
@@ -541,17 +327,17 @@ void zb_osif_serial_recv_data(zb_uint8_t *buf, zb_ushort_t len)
 	/*
 	 * Flush already received data.
 	 */
-	uart_rx_buf_offset = rx_buf_pop(buf, len);
+	uart_rx_buf_offset = ring_buf_get(&rx_ringbuf, buf, len);
```

Comment:
Don't you need to lock the UART irq before this call? The irq might mess things up while you're reading here. There's even a warning in the docs:
```
 * @warning
 * Use cases involving multiple reads of the ring buffer must prevent
 * concurrent read operations, either by preventing all readers from
 * being preempted or by using a mutex to govern reads to the ring buffer.
```

Reply:
It should be protected by the `rx_done_sem` and the fact that the `uart_rx_buf` variable is set after the `ring_buf_get` call retun.

Reply:
I kind of fail to see how it's protected, i. e. what prevents the IRQ from hitting in and taking [this branch](https://github.com/nrfconnect/sdk-nrf/blob/d920a9e119d74cf731c327a48eb3148294802e19/subsys/zigbee/osif/zb_nrf_serial.c#L144), perhaps you could elaborate?

Reply:
You're right. Fixed by disabling the RX interrupt.